lib/Sema/TypeCheckAccess.cpp

```diff
@@ -625,7 +625,21 @@ class AccessControlChecker : public AccessControlCheckerBase,
     if (seenVars.count(theVar) || theVar->isInvalid())
       return;
 
-    checkTypeAccess(theVar->getInterfaceType(), nullptr, theVar,
+    // Get the TypeRepr for this var from its initializer
+    TypeRepr *typeRepr = nullptr;
+    if (auto *E = theVar->getParentInitializer()) {
```

Comment:
This approach seems too specific to the very particular kind of initializer expression used here. I also am not sure it's guaranteed that if you find a `TypeExpr` somewhere in the expression that means that the `NamedPattern`'s type is actually equivalent. For example, what would happen in this test case?

```
internal import Foo

internal func bar(_ x: FooType) -> Bool { return true }

public var y = f(FooType())
```

The type of `y` is `Bool`, not `FooType`, but it seems to me like this code would diagnose `y` as if it were declared to be `FooType`.

I think a better approach to the problem you are trying to solve would be to try to make the diagnostic work even when there is no `TypeRepr`, instead of trying to fabricate a `TypeRepr` for the pattern that does not actually exist.